chat-plugins/calculator.js

```diff
@@ -0,0 +1,111 @@
+'use strict';
+
+function isNumeric(str) {
+	return !isNaN(parseFloat(str));
+}
+
+function parseMathematicalExpression(infix) {
+	// Shunting-yard Algorithm -- https://en.wikipedia.org/wiki/Shunting-yard_algorithm
+	let outputQueue = [];
+	let operatorStack = [];
+	let operators = {
+		"^": {
+			precedence: 4,
+			associativity: "Right",
+		},
+		"/": {
+			precedence: 3,
+			associativity: "Left",
+		},
+		"*": {
+			precedence: 3,
+			associativity: "Left",
+		},
+		"+": {
+			precedence: 2,
+			associativity: "Left",
+		},
+		"-": {
+			precedence: 2,
+			associativity: "Left",
+		},
+	};
+	infix = infix.replace(/\s+/g, "");
+	infix = infix.split(/([+\-*/^()])/);
+	infix = infix.filter(token => token);
+	for (const token of infix) {
+		if ("^*/+-".includes(token)) {
+			let op = operators[token];
+			let prevToken = operatorStack[operatorStack.length - 1];
+			let prevOp = operators[prevToken];
+			while ("^*/+-".includes(prevToken) && (
+				op.associativity === "Left" ? op.precedence <= prevOp.precedence : op.precedence < prevOp.precedence
+			)) {
+				outputQueue.push(operatorStack.pop());
+				prevToken = operatorStack[operatorStack.length - 1];
+				prevOp = operators[prevToken];
+			}
+			operatorStack.push(token);
+		} else if (token === "(") {
+			operatorStack.push(token);
+		} else if (token === ")") {
+			while (operatorStack[operatorStack.length - 1] !== "(") {
+				outputQueue.push(operatorStack.pop());
+			}
+			operatorStack.pop();
+		} else {
+			outputQueue.push(token);
+		}
+	}
+	while (operatorStack.length > 0) {
+		outputQueue.push(operatorStack.pop());
+	}
+	return outputQueue;
+}
+
+function solveRPN(rpn) {
+	let resultStack = [];
+	for (const token of rpn) {
+		if (isNumeric(token) === true) {
+			resultStack.push(token);
+		} else {
+			let a = parseFloat(resultStack.pop());
+			let b = parseFloat(resultStack.pop());
+			switch (token) {
+			case "+":
+				resultStack.push(a + b);
+				break;
+			case "-":
+				resultStack.push(b - a);
+				break;
+			case "*":
+				resultStack.push(a * b);
+				break;
+			case "/":
+				resultStack.push(b / a);
+				break;
+			case "^":
+				resultStack.push(b ** a);
+				break;
+			}
+		}
+	}
+	return resultStack.pop();
+}
+
+exports.commands = {
+	calculate: function (target, room, user) {
+		if (!target) return this.parse('/help calculate');
+		if (!this.runBroadcast()) return;
+		let result = solveRPN(parseMathematicalExpression(target));
+		let validation_regex = /^(?!.*([\^*/+\-.]{2}|\.\d+\.+|^[*/]))[\d+()^*\-/. ]+$/gm;
```

Comment:
Why validate after solving? The validation isn't really strictly necessary, anyway; you'd be filtering out some valid expressions not whitelisted by the validator.

Reply:
Hm, maybe change the condition to:
```
		if (isNumeric(result) === false) {
			return this.errorReply("Something went wrong in the calculation. Maybe check your arithmetical question?");
		} 
```

Reply:
Your `isNumeric` doesn't do that correctly, either. I'd still try to eliminate it.

Reply:
I'd just go with no whitelist at all.

Reply:
I don't think outputting ``NaN`` and ``undefined`` is a good idea tbh

Reply:
> Your isNumeric doesn't do that correctly, either. I'd still try to eliminate it.

what do you mean an answer should _always_ be a number.

Reply:
`1/0 - 1/0` is `NaN`, but I would still just output `NaN` rather than an error message.

Reply:
If you wanted to guard against incorrect syntax like `3*/1`, I would instead throw for things like getting `undefined` from `resultStack.pop()`, and then wrap this part in a try-block.

Reply:
Also feel free to just test `result === undefined`, and return `undefined` for unusual states.